### DISCUSSION TITLE
Pain of 3 hours to configure rust-analyzer and what can be done better

### DISCUSSION THREAD

**Discussion by rusty-snake** [helix-editor/helix#9396] [General]
helix 23.10 (f6021dd0) from Fedora 39 package repositories.

---

I recently switched to helix, rust-analyzer worked out of the box and I am quite happy with it. Now I wanted to configure rust-analyzer to use clippy and other `imports.granularity.group` which got much more complicated than it should be.

[The wiki has an example](https://github.com/helix-editor/helix/wiki/How-to-install-the-default-language-servers#rust) language.toml snippet to use clippy. Unfortunately this snippet does not work. Documentation and Wikis can become outdated so I started debugging this.

Finding the correct format by reading documentation did not worked. Besides the wiki I did not found much documentation how the `config` of a language server is build. And searching for other snippets in the internet and in Issues and Discussions gives more outdated confusion that it helps.

So I had to go with good old trail-and-error. Change `languages.toml` and see what happens. Sounds simple, but it is not because `:config-reload`+`:lsp-restart` can not be used to test although it looks like they are the right commands. `:lsp-restart` restarts rust-analyzer so any change `initilalizationOptions` should be used (this command works as expected). And `:config-reload` re-reads `languages.toml`. You can prove that `languages.toml` is re-read because a syntax error will make `:config-reload` to dump a error message. However, at least anything under `language-server.rust-analyzer.config` is not updated by `:config-reload`. You have to actually restart helix which is very unexpected. After realizing how I can actually test a changed `langauges.toml` after 2hours it got quick.

So back to the original problem. The snippet from the wiki is almost correct, but the stated behaviour of `language-server.rust-analyzer.config.check.command => rust-analyzer.check.command` is wrong. You have to use the full key (i.e. `language-server.rust-analyzer.config.rust-analyzer.check.command`) to make it work.

### What can be done better

- Fix the example in the wiki
- Better document how `language-server.*.config` works. In the best case with a lot of examples in addition to a reference.
- Make `:config-reload` behave in a expected way.

**Comment by kirawi** [comment]
```toml
[language-server.rust-analyzer.config]
check.command = "clippy"
```
does work (I just tested).

I think the reason that it doesn't reload the language config is because `Registry`s reference to the config is never updated to the new one (the old shared pointer is dropped instead of the contents being replaced). That does seem like a bug,
